### PR TITLE
[SYCL] Add new auto device code split mode

### PR DESCRIPTION
**Pull request by AlexeySachkov**
This patch introduces new device code split mode `auto`, which is
intended to automatically select the best device code split mode and
apply it.

At the moment, `auto` is equivalent to `per_source` for most cases and
it is equivalent to `off` in case of precense of function pointers.

### CODE REVIEW

**Comment by AGindinson** [comment]
I would argue that the current title is somewhat misleading: before revising the actual changes, I've mistaken this for "always run the split unless `-fsycl-device-code-split=off` is specified, even if no option was passed at all". Maybe something like:
```
[SYCL] Enable a default mode for device code split
```

**Review by Fznamznon — COMMENTED** [review]
Let's also add tests.

**Review by AGindinson — DISMISSED** [review]
The driver part LGTM.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 1]
```diff
@@ -290,6 +292,25 @@ enum KernelMapEntryScope {
...
+  // is the best in each particular situation.
+  // At the moment, we assume that per-kernel split is the best way of splitting
+  // device code and it can be always selected unless there are functions marked
+  // with [[intel::device_indirectly_callable]] attribute, because it instructs
+  // us to make this function available to the whole program as it was compiled
+  // as a single module.
+  bool HasDeviceIndirectlyCallable = false;
+  for (auto &F : M.functions()) {
+    if (F.hasFnAttribute("referenced-indirectly"))
+      HasDeviceIndirectlyCallable = true;
```
[the remark this answers is not in the document]
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Fixed in 417593249c45af36535c11e142e75d3219190601

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 2]
```diff
@@ -290,6 +292,25 @@ enum KernelMapEntryScope {
   Scope_Global     // single entry in the map for all kernels
 };
 
+static KernelMapEntryScope selectDeviceCodeSplitModeAutomatically(Module &M) {
```
[the remark this answers is not in the document]
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Applied in 417593249c45af36535c11e142e75d3219190601

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 3]
```diff
@@ -90,16 +90,18 @@ static cl::opt<bool> OutputAssembly{"S",
...
 enum IRSplitMode {
   SPLIT_PER_TU,    // one module per translation unit
-  SPLIT_PER_KERNEL // one module per kernel
+  SPLIT_PER_KERNEL, // one module per kernel
+  SPLIT_AUTO // automatically select split mode
 };
 
 static cl::opt<IRSplitMode> SplitMode(
     "split", cl::desc("split input module"), cl::Optional,
     cl::init(SPLIT_PER_TU),
```
Comment (kbobrovs) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:99]:
```suggestion
    cl::init(SPLIT_AUTO),
```

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 4]
```diff
@@ -289,6 +291,42 @@ enum KernelMapEntryScope {
   Scope_Global     // single entry in the map for all kernels
 };
 
+static KernelMapEntryScope selectDeviceCodeSplitScopeAutomatically(Module &M) {
+  // Here we can employ various heuristics to decide which way to split kernels
```
Comment (kbobrovs) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:295]: There are more comments then the code here :) I think the idea is pretty straightforward and the description could be shrinked.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 5]
```diff
@@ -624,9 +662,9 @@ int main(int argc, char **argv) {
     errs() << "no actions specified; try --help for usage info\n";
     return 1;
   }
-  if (IROutputOnly && DoSplit) {
-    errs() << "error: -" << SplitMode.ArgStr << " can't be used with -"
-           << IROutputOnly.ArgStr << "\n";
+  if (IROutputOnly && (DoSplit && SplitMode != SPLIT_AUTO)) {
```
Comment (kbobrovs) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:665]: Why SPLIT_AUTO is allowed with IR output only?

Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:665]: > Why SPLIT_AUTO is allowed with IR output only?

Technically this is me being a bit lazy to dig further into the driver and do some changes there. Basically, there are compilation flows (related to AOT), where `sycl-post-link` is involved with `-ir-output-only` option and according to my changes, `-split=` is almost always passed to `sycl-post-link` (unless split was explicitly disabled) and it triggered errors in some LIT tests.

Comment (kbobrovs) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:665]: OK, so the `auto` heuristics in presence of `-ir-output-only` is fixed to "don't split" - ? please add a comment then, to user-visible description too

File: llvm/tools/sycl-post-link/sycl-post-link.cpp (hunk 3 above)
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Changed in edfa44cc9a08c732d48f3bfe2b8caf3f8841d4ac

File: llvm/tools/sycl-post-link/sycl-post-link.cpp (hunk 4 above)
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: I've re-structured comments a bit in edfa44cc9a08c732d48f3bfe2b8caf3f8841d4ac

File: llvm/tools/sycl-post-link/sycl-post-link.cpp (hunk 5 above)
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:667]: > OK, so the auto heuristics in presence of -ir-output-only is fixed to "don't split" - ? please add a comment then, to user-visible description too

That's correct, updated help message of the tool in edfa44cc9a08c732d48f3bfe2b8caf3f8841d4ac

File: clang/include/clang/Driver/Options.td [hunk 6]
```diff
@@ -1987,11 +1987,11 @@ def fsycl_link_targets_EQ : CommaJoined<["-"], "fsycl-link-targets=">, Flags<[No
...
    Flags<[CC1Option, CoreOption]>, HelpText<"Perform SYCL device code split: per_kernel (device code module is "
-  "created for each SYCL kernel) | per_source (device code module is created for each source (translation unit)) | off (no device code split). "
-  "Default is 'off' - all kernels go into a single module`">, Values<"per_source, per_kernel, off">;
+  "created for each SYCL kernel) | per_source (device code module is created for each source (translation unit)) | off (no device code split). | auto (use heuristic to select the best way of splitting device code)"
+  "Default is 'auto' - automatically select how to split device code into modules">, Values<"per_source, per_kernel, off, auto">;
 def fsycl_device_code_split : Flag<["-"], "fsycl-device-code-split">, Alias<fsycl_device_code_split_EQ>,
-  AliasArgs<["per_source"]>, Flags<[CC1Option, CoreOption]>,
-  HelpText<"Perform SYCL device code split in the per_source mode i.e. create a device code module for each source (translation unit)">;
+  AliasArgs<["auto"]>, Flags<[CC1Option, CoreOption]>,
+  HelpText<"Perform SYCL device code split in the 'auto' mode i.e. use heuristic to distribute device code across modules">;
```
Comment (AGindinson) [inline clang/include/clang/Driver/Options.td:1994]:
```suggestion
  HelpText<"Perform SYCL device code split in the 'auto' mode, i.e. use heuristic to distribute device code across modules">;
```

File: clang/include/clang/Driver/Options.td [hunk 7]
```diff
@@ -1987,11 +1987,11 @@ def fsycl_link_targets_EQ : CommaJoined<["-"], "fsycl-link-targets=">, Flags<[No
   HelpText<"Specify comma-separated list of triples SYCL offloading targets to produce linked device images">;
 def fsycl_device_code_split_EQ : Joined<["-"], "fsycl-device-code-split=">,
    Flags<[CC1Option, CoreOption]>, HelpText<"Perform SYCL device code split: per_kernel (device code module is "
-  "created for each SYCL kernel) | per_source (device code module is created for each source (translation unit)) | off (no device code split). "
-  "Default is 'off' - all kernels go into a single module`">, Values<"per_source, per_kernel, off">;
+  "created for each SYCL kernel) | per_source (device code module is created for each source (translation unit)) | off (no device code split). | auto (use heuristic to select the best way of splitting device code)"
+  "Default is 'auto' - automatically select how to split device code into modules">, Values<"per_source, per_kernel, off, auto">;
```
Comment (AGindinson) [inline clang/include/clang/Driver/Options.td:1991]: Minor: can this be unified with the alias description?
```suggestion
  "Default is 'auto' - use heuristic to distribute device code across modules">, Values<"per_source, per_kernel, off, auto">;
```

**Review by AGindinson — APPROVED** [review]
The driver part LGTM.

**Review by kbobrovs — APPROVED** [review]

**Pull request merged**